README.md
<!--
 * @Description: 
 * @Date: 2022-04-14 15:28:56
 * @LastEditTime: 2022-07-11 16:34:14
-->

# electron 静默打印

项目基于 [electron-vue](https://github.com/SimulatedGREG/electron-vue) 基础上开发, 使用的 electron 版本为 **11.2.1** 。

> 使用版本: node: 14.16.0 ,  npm: 6.14.11 , yarn: 1.22.10
> 
> >  [electron-vue中文指引](https://simulatedgreg.gitbooks.io/electron-vue/content/cn/)

## 使用

```bash
# 安装
yarn install 

# 运行
yarn dev

## win32 打包
yarn build:win32

## win64 打包
yarn build
```

## 解决问题

1. 前端可以直接调本地服务接口，将文件进行静默打印，不需要弹窗提示 ( 类似 window.print 的弹窗 )；
2. 打印进度提示与打印队列；

## 实现思路

1. 应用打开时默认启动  express 服务；
2. 前端调用应用启动的 express 服务接口，将文件路径作为参数传过去；
3. 应用接口收到内容后，包装一层数据，将其传进写好的打印队列调度器中进行打印任务处理；
4. 打印成功或失败会将内容从接口中返回到前端；

## 前端调用

> 详细请看 example/test-print.html

```javascript
// 应用默认启动 express 的端口为 45656

// 单个打印接口
const url = 'http://localhost:45656/print'
const fileUrl = 'http://xxxxxxxxx' // 文件地址
fetch(`${url}?fileUrl=${fileUrl}`, { method: 'POST' })
// 或
fetch(`${url}`, { method: 'POST', body: { fileUrl: fileUrl, downloadOptions: { xxx } } })

// 批量打印接口
const url = 'http://localhost:45656/multiple-print'
let fileUrls = ['http://aaaa', 'http://vvvv', 'http://dddd']
fileUrls = fileUrls.join(';') // 需要用 ';' 拼接起来
fetch(`${url}?fileUrl=${fileUrl}`, { method: 'POST' })

// 预览接口
const url = 'http://localhost:45656/preview'
const fileUrl = 'http://yyyyyyy'
fetch(`${url}?fileUrl=${fileUrl}`)
```

## 已实现功能

1. app 右下角托盘显示；
2. app 开机自启动；
3. 屏幕右下角打印进度条提示；
4. 通过自定义协议打开/唤醒 app
5. 单个打印、批量打印、打印预览；

## PDF打印过程

1. 将文件下载到本地缓存文件夹中( 使用 [download](https://github.com/kevva/download) );
2. 将下载好的本地文件打印 ( 使用 [pdf-to-printer](https://github.com/artiebits/pdf-to-printer) )；
3. 打印完成删除文件；

## 注意

1. 尚未对预留的 socket 内容进行开发，目前仅支持 http 请求;
2. 项目基于win平台开发，mac平台上部分功能未做兼容处理;

## 支持打印的文件类型

1. PDF (.pdf) **pdf打印目前仅支持windows平台**;
2. 图片 (.jpg .png .jpeg);

## 第三方包推荐

1. 日志调试 ( [electron-log](https://github.com/megahertz/electron-log) );
   electron-log 日志所在位置: 
   - **on Linux:** ~/.config/{app name}/logs/{process type}.log
   - **on macOS:** ~/Library/Logs/{app name}/{process type}.log
   - **on Windows:** %USERPROFILE%\AppData\Roaming{app name}\logs{process type}.log
2. 自动升级 ( electron-updater );

ChangeLog.md
<!--
 * @Description: change logs
 * @Date: 2022-04-14 15:49:34
 * @LastEditTime: 2023-05-24 10:54:46
-->

# 更新记录

## 2023/05/15

> version 0.3.3

1. windows 开机自启动时, 默认不显示主窗口；

## 2023/05/08

> version 0.3.1

1. 下载pdf时, 增加下载超时/失败的错误提示；

## 2023/04/17

> version 0.3.0

1. 使用`download`下载文件时，增加 **options** 选项，用于`download`的入参（目前仅适配了单个打印）

## 2023/03/16

> version 0.2.4

1. 增加"关于"页描述app信息；
2. 修复默认打开预览窗口时，首次闪烁的问题；

## 2022/07/11

> version 0.2.3

1. 修复打印完成后未删除cache的问题;
2. 获取文件后缀名时进行转小写处理;

## 2022/07/06

> version 0.2.3

1. 修复下载文件后，删除缓存失败的问题；

## 2022/04/14

> version 0.1.3

1. 新增自定义协议唤醒 `app` 功能;
2. 协议触发时中可以通过设置 `hide` 字段来显示唤醒存在的实例时是否需要显示隐藏主界面；


static/express-process/app.js
/* eslint-disable no-unused-vars */
/*
 * @Description: express 单个打印服务
 * @Date: 2022-01-10 17:29:37
 * @LastEditTime: 2023-05-08 15:46:42
 */
const express = require('express')
const bodyParser = require('body-parser')
const PrintScheduler = require('./print/scheduler.js')
const { handleFileType } = require('./print/file-type')

const app = express()

const { downloadFile, printPdf, deleteCache, wrapRandomFolder, printImage } = require('./print/index.js')

const [,, __static, httpPort = 45656, socketPort, cacheDir, deviceName] = process.argv
console.log(__static, httpPort, socketPort, cacheDir, deviceName)

// NOTE: 打印队列实例
const printScheduler = new PrintScheduler(handlePrint)

// eslint-disable-next-line no-constant-condition
if (10 > 11) {
  console.log(socketPort, deviceName)
  console.log(printPdf)
}

let fileUid = 0 // 每个文件打印的 id 标识 (唯一, 会递增)
let multipleQueueIndex = 0 // 同一个打印队列的 id 标识 (唯一, 会递增)

// 允许跨域
app.all('*', (req, res, next) => {
  // console.log(req.headers.origin)
  // console.log(req.environ)
  res.header('Access-Control-Allow-Origin', req.headers.origin)
  // res.header("Access-Control-Allow-Origin", '*');
  res.header('Access-Control-Allow-Headers', 'Content-Type,Content-Length, Authorization, Accept,X-Requested-With')
  res.header('Access-Control-Allow-Methods', 'PUT,POST,GET,DELETE,OPTIONS')
  res.header('Access-Control-Allow-Credentials', 'true')
  res.header('X-Powered-By', ' 3.2.1')
  if (req.method === 'OPTIONS') res.send(200)/* 让options请求快速返回 */
  else next()
})

// NOTE: 在路由处理程序之前添加 body-parser 中间件
app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.get('/', (req, res) => {
  res.send({
    code: '200',
    message: 'express-print'
  })
})

/**
 * 预览文件 (只能单个预览)
 */
app.get('/preview', (req, res) => {
  const { fileUrl } = req.query
  console.log('preview - ', fileUrl)

  if (!fileUrl) res.send({ code: '400', message: 'fileUrl不能为空' })
  else {
    // 通知父进程, 然后通过  EventEmitter 来触发 file-preview.js 中的创建窗口预览
    process.send({ type: 'preview-file', result: fileUrl })
    res.send(204)
  }
})

/**
 * 多个打印
 * NOTE: 用 ; 隔开
 * 事例: localhost:45656/print?fileUrl=http://xxxx;http://zzzzzz;http://yyyyy
 */
app.post('/multiple-print', (req, res) => {
  try {
    const { fileUrl } = req.query
    const urls = fileUrl.split(';')
    const wrapUrls = urls.map(url => {
      const printInfo = createPrintInfo(url, false)
      return printInfo
    })

    const startUid = wrapUrls[0].__fileUid__ // 该段队列的起始点 fileUid
    const endUid = wrapUrls[wrapUrls.length - 1].__fileUid__ // 该段队列的结束点 fileUid

    const printCallback = (status, data) => {
      let respondData = {}
      if (status === 1) {
        respondData = {
          code: '200',
          data,
          message: `打印完成`
        }
      }
      res.send(respondData)
    }

    wrapUrls.forEach((element, index) => {
      // ==========
      element.start = startUid
      element.end = endUid
      element.__multipleId__ = multipleQueueIndex // 每个批量的队列都有自己的 id 标识
      // ==========

      printScheduler.insert(element, printCallback)

      // 该条多个打印队列中添加回调
      if (wrapUrls.length - 1 === index) {
        element.successCallback = printCallback // 将队列最后一个成员添加 callback
      }
    })

    multipleQueueIndex++
  } catch (e) {
    res.send({
      code: '400',
      message: e + ''
    })
  }
})

/**
 * 单个打印服务
 * NOTE: 需要创建打印队列, 因为打印机只能一份一份打印
 *
 * 事例: localhost:45656/print?fileUrl=http://xxxx
 */
app.post('/print', (req, res, next) => {
  console.log('-------------------------------------------')
  /**
   * NOTE:
   * 实现步骤:
   * 1. 在目标缓存目录下继续创建随机文件夹(目的防止文件重名覆盖或重命名)
   * 2. 通过步骤 1 在目标文件夹下下载文件
   * 3. 下载完成后打印对应的 pdf 文件
   * 4. 打印完删除对应的随机文件夹 (不是删除文件)
   */
  const { fileUrl: fileUrlInUrl } = req.query

  // WARM: 当前版本设计, fileUrl 先不从 url 中移到 body(保留在 url 中) , 因为这个调用属于破坏性更新, 将影响之前版本的调阅方法
  // 目前才用合并处理, 如果 url 中存在 fileUrl 中的话, 就采用请求 url 中的 fileUrl (反之取 body 中的)

  const { downloadOptions, fileUrl: fileUrlInBody } = req.body

  const fileUrl = fileUrlInUrl || fileUrlInBody

  console.log('文件地址: ', fileUrl)

  let respondData = {}
  if (fileUrl) {
    const printInfo = createPrintInfo(fileUrl, true, downloadOptions)

    const printCallback = (status = 1, e) => {
      // 成功
      if (status === 1) {
        respondData = {
          code: '200',
          message: `打印完成`
        }
      } else {
        respondData = {
          code: '500',
          message: `${e}`
        }
      }

      res.send(respondData)
    }

    printScheduler.insert(printInfo, printCallback)
  } else {
    respondData = {
      code: '400',
      message: 'fileUrl不能为空'
    }
    res.send(respondData)
  }

  // next()
})

/**
 * 创建打印任务信息
 * @param {String} fileUrl - 文件地址
 * @param {Boolean} isSingle - 是否单个打印
 * @param {object} downloadOptions - download 下载库的 options 选项
 *
 * @returns {Object}
 */
function createPrintInfo (fileUrl, isSingle = true, downloadOptions = {}) {
  if (!fileUrl) throw new Error('fileUrl不能为空')
  const printInfo = {
    cacheDir,
    fileUrl,
    __fileUid__: fileUid,
    isSingle,
    downloadOptions
  }
  fileUid++

  return printInfo
}

/**
 * 处理打印任务
 * @returns {Promise} { status:Boolean, ?error:String }
 */
async function handlePrint ({ fileUrl, downloadOptions }) {
  /**
   * 是否打印成功
   */
  let status = false

  /**
   * 打印失败时的错误信息
   */
  let error = ''

  /**
   * 包装一下文件夹 - 让它变成随机的
   */
  let randomCacheDir = ''

  /**
   * 真实文件名
   */
  let realFilename = ''

  try {
    randomCacheDir = wrapRandomFolder(cacheDir)

    // 下载文件至本地的缓存文件夹
    const { filename, fileType } = await downloadFile(
      fileUrl,
      randomCacheDir,
      downloadOptions
    )

    // // 没有文件后缀的, 要处理, 因为这有可能涉及到资源权限, 有的资源可能做了 授权, 防盗链 等处理。可能是个 http 请求
    // if (fileType) {}

    realFilename = filename

    // NOTE: 正式打印 ========== ↓
    await dispatchByFileType(
      handleFileType(fileType),
      filename,
      randomCacheDir,
      fileUrl
    )
    // NOTE: 正式打印 ========== ↑

    status = true
  } catch (e) {
    console.log(e, '\n--------------------')

    status = false
    error = (e?.message || e) + ''
  } finally {
    // 无论是否打印成功, 都删除已下载的文件(直接删除文件夹)
    deleteCache(randomCacheDir)

    console.log(`===== 是否打印成功: ${status ? '是' : '否'}`, realFilename, '=====')
  }

  return { status, error }
}

/**
 * 根据不同的文件来处理打印
 * 依据 handleFileType 的返回值
 * @param {string} fileDetail - 文件具体类型
 * @param {string} filename - 文件夹
 * @param {string} randomCacheDir - 通过随机生成的文件夹名字
 * @param {string} fileUrl - 资源地址
 */
function dispatchByFileType (fileDetail, filename, randomCacheDir, fileUrl) {
  // 这一步, 使用 download 下载文件后, 需要看文件下载的内容是否支持打印
  // NOTE: 目的: 为了对不同的文件类型使用不同的打印方式进行分发处理

  switch (fileDetail) {
    case 'image':
      // 打印图片

      return printImage(randomCacheDir, filename, deviceName)

    case 'pdf':
      // 打印 pdf

      return printPdf(randomCacheDir, filename, deviceName)

    default :
      // 未匹配

      throw Error('未支持打印该文件类型, 请检查链接资源是否正确且可以被访问')
  }
}

// 启动服务
app.listen(Number(httpPort), () => void console.log(`server open on ${Number(httpPort)}!`))
